posts/inside-rust/2023-05-02-stabilizing-async-fn-in-trait.md

```diff
@@ -0,0 +1,197 @@
+
+[RFC 3185]: https://rust-lang.github.io/rfcs/3185-static-async-fn-in-trait.html
+
+
+## MVP Part 2: Send bounds and associated return types
+
+There is one complication that arises when using async functions in traits that doesn't arise with sync functions. Many async runtimes -- notably including the default configurations of [Tokio] and [async-std] -- use a workstealing thread scheduler. This means that futures may move between worker threads dynamically to achieve load balancing. As a result, the future must only capture `Send` data.
+
+[Tokio]: https://tokio.rs/
+
+[async-std]: https://async.rs/
+
+If you author a generic async function that spawns tasks on one of those runtimes, however, you will start to get compilation errors ([playground](XXX)):
+
+```rust
+async fn do_health_check_par(hc: impl HealthCheck) {
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+The problem is that the future returned by `hc.check()` isn't guaranteed to be `Send`. It might access non-Send data. The solution is to add a `Send` bound, but given that this is an async function, it's not obvious how to do that. How do we talk about the future returned by a call to `hc.check()`? Associated return types provide the answer. We can convert the above function to use an explicit type parameter `HC` (instead of `impl HealthCheck`) and then add a new bound, `HC::check(): Send`. This says "the value returned by `HC::check` must be of `Send` type":
+
+```rust
+async fn do_health_check_par<HC>(hc: HC)
+where
+    HC: HealthCheck + Send + 'static,
+    HC::check(): Send, // <-- associated return type
+{
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+Of course, it's kind of unfortunate that we had to rewrite from taking an `impl HealthCheck` to an explicit `HC` type parameter in order to use this notation. RFC #2289, "associated type bounds", introduced a compact notation to address this problem. That RFC is not part of this MVP, but if it were stabilized, then one could simply write:
+
+```rust
+async fn do_health_check_par(hc: impl HealthCheck<check(): Send> + Send + 'static) {
+    //                                            -------------
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+In our [previous post][pp], we [hypothesized](https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html#hypothesis-this-is-uncommon) that this problem might not occur often in practice. However, our case studies found that it comes up quite frequently, and so we decided that a solution is needed. We explored a number of solutions and concluded that associated return types are the most practical.
+ 
```

Comment:
I think we should probably have async methods be Send by default and use RTN to opt out of the Send-ness (given that Send futures are more common). We should discuss this at some point before getting too far along with stabilisation. Not sure it's worth mentioning the possibility in the blog post?

Reply:
We have discussed this on a number of occasions already, and the following two objections are usually raised:
1. Rust uses opt-in `Send` bounds everywhere; making `async fn` in traits imply `+ Send` by default would be inconsistent with the rest of the language. For example: `async fn` free functions do not imply `+ Send`, making the "send"-ness of `async fn` contextually dependent on whether it's used in a trait. That seems tricky to remember, and easy to mess up.
2. While `+ Send` types and bounds are the norm today for `async-std`/`tokio`-style executors today, these aren't the only domain and executor styles we need to care about. My understanding is that this would cause issues for single-threaded domains such as embedded. @dirbaio should be able to elaborate this point.

Reply:
Yes, in embedded the default is single-threaded executors. Multi-core chips are rare, and most of them are "asymmetric" (different cores have separate dedicated RAM/flash, separate peripherals, different speeds or even different architecture). Multi-core is usually used by manually assigning different functions to different cores, to get better isolation/security/predictability thanks to not having latency for one function impacted by other functions running on the same core. You never need to "load-balance work" between multiple cores to scale performance, these chips are usually slow (<200Mhz) so it's simpler to get faster cores if you need more perf. It's only when you reach multi-GHz territory that you need multiple cores to scale perf.

So, embedded uses one single-threaded executor for each core, instead of a big multithreaded one. Having AFIT be Sync by default would be very, very annoying, and inconsistent with the rest of the language. Forcing Send is especially annoying in embedded, because you don't have `std::sync::Mutex`, you often have no heap and some archs dont't have atomic CAS (like `thumbv6`, so no `Arc` either.

Reply:
> We have discussed this on a number of occasions already, and the following two objections are usually raised:

We have, but I feel like every time the discussion has just kind of drifted off without us making a plan or coming to a decision (in part, because decisions about defaults are hard - its just about ergonomics rather than anything which can be argued logically). I don't really want to discuss here, my comment was just about whether we should say something about it in the blog post.

To address your points though, Rust doesn't use opt-in Send bounds, it infers the Send bounds. It's just that with trait methods you can't infer them so we have to pick a default. This will be confusing either way, so we just have to pick the least confusing. Which segues into point 2 - nobody is saying we shouldn't support non-Send methods at all, just whether they should be opt-in or opt-out. Single-threaded runtimes are totally valid and should be supported, but I don't see them being more popular than work stealing. To clarify, these are defaults for the callsite, not for the declaration, so I don't think there is any downside other than more typing.

Reply:
We should move this discussion to Zulip.